File: server/src/utils/cache.js
// Simple in-memory cache with TTL

class Cache {
    constructor(ttlMs = 5 * 60 * 1000) { // Default 5 minutes
        this.cache = new Map();
        this.ttlMs = ttlMs;
    }

    /**
     * Get a value from cache
     * @param {string} key
     * @returns {any|null}
     */
    get(key) {
        const entry = this.cache.get(key);
        if (!entry) return null;

        if (Date.now() > entry.expiresAt) {
            this.cache.delete(key);
            return null;
        }

        return entry.value;
    }

    /**
     * Set a value in cache
     * @param {string} key
     * @param {any} value
     * @param {number} ttlMs - Optional custom TTL
     */
    set(key, value, ttlMs = this.ttlMs) {
        this.cache.set(key, {
            value,
            expiresAt: Date.now() + ttlMs
        });
    }

    /**
     * Check if key exists and is not expired
     * @param {string} key
     * @returns {boolean}
     */
    has(key) {
        return this.get(key) !== null;
    }

    /**
     * Clear all cache entries
     */
    clear() {
        this.cache.clear();
    }

    /**
     * Get cache stats
     */
    stats() {
        return {
            size: this.cache.size,
            ttlMs: this.ttlMs
        };
    }
}

// Export singleton instance for price caching
export const priceCache = new Cache(5 * 60 * 1000); // 5 minute TTL

export default Cache;

File: server/app/utils/cache.py
"""
Simple TTL cache for Universalis API responses.
"""

from cachetools import TTLCache
from typing import Any

# 5-minute TTL cache for price data
price_cache: TTLCache = TTLCache(maxsize=1000, ttl=300)  # 300 seconds = 5 minutes


def get_cached(key: str) -> Any | None:
    """Get a value from cache."""
    return price_cache.get(key)


def set_cached(key: str, value: Any) -> None:
    """Set a value in cache."""
    price_cache[key] = value


def has_cached(key: str) -> bool:
    """Check if key exists in cache."""
    return key in price_cache


def cache_stats() -> dict:
    """Get cache statistics."""
    return {
        "size": len(price_cache),
        "maxsize": price_cache.maxsize,
        "ttl": price_cache.ttl
    }
